python_modules/dagster/dagster/_core/definitions/definitions_fn.py

```diff
@@ -0,0 +1,124 @@
+import inspect
+import itertools
+import sys
+from contextlib import contextmanager
+from types import ModuleType
+from typing import Any, Iterable, Mapping, Optional, Union
+
+from dagster import _check as check
+from dagster._core.execution.with_resources import with_resources
+
+from .assets import AssetsDefinition, SourceAsset
+from .decorators import repository
+from .job_definition import JobDefinition
+from .repository_definition import PendingRepositoryDefinition, RepositoryDefinition
+from .resource_definition import ResourceDefinition
+from .schedule_definition import ScheduleDefinition
+from .sensor_definition import SensorDefinition
+
+MAGIC_REPO_GLOBAL_KEY = "__dagster_repository"
+
+NO_STACK_FRAME_ERROR_MSG = "Python interpreter must support Python stack frames. Python interpreters that are not CPython do not necessarily implement the necessary APIs."
+
+# invoke this function to get the module name the function that called the current
+# scope
+def get_module_name_of_caller() -> str:
+    # based on https://stackoverflow.com/questions/2000861/retrieve-module-object-from-stack-frame
+    # two f_backs to get past get_module_name_of_caller frame
+
+    # Need to do none checking because of:
+    # https://docs.python.org/3/library/inspect.html#inspect.currentframe
+    # CPython implementation detail: This function relies on Python stack frame
+    # support in the interpreter, which isn’t guaranteed to exist in all
+    # implementations of Python. If running in an implementation without
+    # Python stack frame support this function returns None.
+
+    frame = check.not_none(inspect.currentframe(), NO_STACK_FRAME_ERROR_MSG)
+    back_frame = check.not_none(frame.f_back, NO_STACK_FRAME_ERROR_MSG)
+    back_back_frame = check.not_none(back_frame.f_back, NO_STACK_FRAME_ERROR_MSG)
+    return back_back_frame.f_globals["__name__"]
+    # without error checking
+    # return inspect.currentframe().f_back.f_back.f_globals["__name__"]
+
+
+def get_python_env_global_dagster_repository() -> RepositoryDefinition:
+    parent_module_name = get_module_name_of_caller()
+    return sys.modules[parent_module_name].__dict__[MAGIC_REPO_GLOBAL_KEY]
+
+
+@contextmanager
+def definitions_test_scope(dundername):
+    parent_mod = sys.modules[dundername]
+    assert MAGIC_REPO_GLOBAL_KEY not in parent_mod.__dict__
+    try:
+        yield
+    finally:
+        if MAGIC_REPO_GLOBAL_KEY in parent_mod.__dict__:
+            del parent_mod.__dict__[MAGIC_REPO_GLOBAL_KEY]
+
+
+class DefinitionsAlreadyCalledError(Exception):
+    pass
+
+
+def get_dagster_definitions_in_module(mod: ModuleType):
+    return mod.__dict__[MAGIC_REPO_GLOBAL_KEY]
+
+
+# TODO: Add a new Definitions class to wrap RepositoryDefinition?
+def definitions(
```

Comment:
do we want to consider adding some sugar for `load_all_assets_from_package_name`? i.e. an arg `asset_package_names: List[str]`?

Reply:
I believe we should leave that to user to compose in this API. Otherwise we will end up supporting a large surface area of potential ways to support asset loading in this top-level API.